sklearn/neighbors/base.py

```diff
@@ -439,11 +439,12 @@ class from an array representing our data set and ask who's
                 raise ValueError(
                     "%s does not work with sparse matrices. Densify the data, "
                     "or set algorithm='brute'" % self._fit_method)
-            if (sys.version_info < (3,) or
-                    LooseVersion(joblib_version) < LooseVersion('0.12')):
+            old_joblib = LooseVersion(joblib_version) < LooseVersion('0.12')
+            if sys.version_info < (3,) or old_joblib:
                 # Deal with change of API in joblib
+                check_pickle = False if old_joblib else None
                 delayed_query = delayed(_tree_query_parallel_helper,
-                                        check_pickle=False)
+                                        check_pickle=check_pickle)
```

Comment:
I am not sure why we have a different behavior with `python2.7` here.
Shouldn't is be only dependent on the version of `joblib`?

Reply:
You introduced this in #12172 because in 2.7 balltree is not picklable (according to your issue description)

Reply:
I did not realize I was to blame for this. Yes indeed it is needed.
Sorry for the noise.